proposals/collection-expressions-inline.md

```diff
@@ -0,0 +1,124 @@
+
+> An *input type inference* is made *from* an expression `E` *to* a type `T` in the following way:
+>
+> * If `E` is a *collection expression* with elements `Eᵢ`, and `T` is a type with an [*iteration type*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/statements.md#1295-the-foreach-statement) `Tₑ` or `T` is a *nullable value type* `T0?` and `T0` has an [*iteration type*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/statements.md#1295-the-foreach-statement) `Tₑ`, then for each `Eᵢ`:
+>   * If `Eᵢ` is an *expression element*, then an *input type inference* is made *from* `Eᵢ` *to* `Tₑ`.
+>   * If `Eᵢ` is a *spread element* `..Sᵢ`, then a [*lower-bound inference*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#116310-lower-bound-inferences) is made *from*  the [*iteration type*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/statements.md#1295-the-foreach-statement) of `Sᵢ` *to* `Tₑ`.
+> * *[existing rules from first phase]* ...
+
+### Foreach
+
+The collection in a `foreach` statement may be a *collection expression*.
+
+If the `foreach` statement has an *explicitly typed iteration variable* of type `Tₑ`, the compiler verifies the following for each element `Eᵢ` in the collection expression and reports an error otherwise:
+* If `Eᵢ` is an *expression element*, there is an implicit conversion from `Eᵢ` to `Tₑ`.
+* If `Eᵢ` is a *spread element* `..Sᵢ`, there is an implicit conversion from `Sᵢ` to the *collection_type* `col<Tₑ>`.
+
+If the `foreach` statement has an *implicitly typed iteration variable*, the type of the *iteration variable* is the [*best common type*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#116315-finding-the-best-common-type-of-a-set-of-expressions) of the types of the elements. If there is no *best common type*, an error is reported.
+
+For each element `Eᵢ` in the collection expression, the type if any contributing to the *best common type* is the following:
+* If `Eᵢ` is an *expression element*, the type of `Eᵢ`.
+* If `Eᵢ` is a *spread element* `..Sᵢ`, the [*iteration type*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/statements.md#1295-the-foreach-statement) of `Sᵢ`.
+
+Since collection expressions do not have a *type* or an *iteration type*, the *best common type* does not consider elements from nested collection expressions.
+
+For a collection expression used as the collection in a `foreach` statement, the compiler may use any conforming representation for the collection instance, including eliding the collection.
+
+The elements of the collection are evaluated in order, and the loop body is executed for each element in order. The compiler *may* evaluate subsequent elements before executing the loop body for preceding elements.
```

Comment:
> The compiler *may* evaluate subsequent elements before executing the loop body for preceding elements.

Is that really something that should be left unspecified? The compiler cannot easilly change it later, as that would be an observable breaking change, and one that feels pretty likely to break real programs, whenever somebody does `foreach (Foo foo in [myFoo, ..someFunctionWithSideEffects()])`, especially if someFunctionWithSideEffects might mutate `myFoo`.

I'd honestly expect all the elements to be evaluated first before any loop executions, simply because that it what would happen if the expression were converted to *literally any* runtime type.

The only implementation strategies I can even think of that would allow for delayed evaluation would be unrolling the loop body, or the creation of a generator that yields the elements one by one, and then iterating over that generator's output.

Reply:
> Is that really something that should be left unspecified? 

Yes.  It allows the compiler great leeway in the impl strategy for converting the expressions to a collection.

> The compiler cannot easilly change it later, as that would be an observable breaking change

It's not a breaking change as it's explicitly marked as unspecified.  Taking a dependency on it puts one in unsupported territory, with changes free to happen.

> and one that feels pretty likely to break real programs,

I would definitely hope not.

> whenever somebody does foreach (Foo foo in [myFoo, ..someFunctionWithSideEffects()]), especially if someFunctionWithSideEffects might mutate myFoo.

NOte: it's fine for `someFunctionWithSideEffects` to have side effects.  Those will happen *after* `myFoo` is evaluated, and before any other elements are evaluated.  What is unspecified is what it means is if the *iteration* of that evaluated value itself mutates.  That would be a pattern we consider too 'out there', and we're willing to say that if your iterators do this, then collection expressions are not a good fit for you.  We want to optimzie for the 99.99% case, where iteration order will not have an impact. 

Note: to reiterate.  Evaluation-order of the *elements* is still guaranteed to be left-to-right, just as it always has been in C#.  So side effects from prior element evaluations will be seen by later element evaluations.

> I'd honestly expect all the elements to be evaluated first before any loop executions

That would require stack space for all their values.  That would def be undesirable in many cases.  Simply imagine a collection expr with 1000 elements.  If we require all that evaluation up front, we need to store *all* those values somewhere (like teh stack) then copy them into the dest.  However, if we give the compiler freedom, it can do something like use a single space on the stack, evaluate each element it it, add to the dest, and the move on to the next (reusing that single space).

Reply:
>>and one that feels pretty likely to break real programs,
>I would definitely hope not.

A better example is if evaluation of any element throws. I can predict with high certainty that users will notice and care about the difference between processing the loop body for the first handful elements and then throwing vs throwing before processing any elements. Imagine if say the loop body is writing to a file and the 15th element throws when evaluated. Pretty significant difference if the file ends up empty, vs if it has lines for the previously processed elements. 

If initially implemented with one at a time evaluation, then users may come to depend on the file having the lines from all elements before the throwing was was evaluated. If initially pre-evaluating all elements then users could easily become reliant on that behavior. 

And files are hardly the only such scenarios. This could also apply to GUI scenarios like adding elements to listbox, or whatever. 

Obviously users understand they can get partial execution of the loop if the body throws. And users quickly learn about lazy Enumerables, like those created by generators and or LINQ, where they can get partial execution from the enumerator throwing. But users are used to any type of foreach-able construct being eagerly evaluated (arrays and most collection classes) or lazily evaluated.

But here you are trying to reserve the right for the compiler to sometimes treat the collection expression as eagerly evaluated (like arrays or spans), and other times to be lazily evaluated like a generator or LINQ Enumerables, and possibly even mix the two within one expression in undefined ways that may change in the future.  Which is very much not something users are used to, and something which will be hard to explain to users.

>It's not a breaking change as it's explicitly marked as unspecified. Taking a dependency on it puts one in unsupported territory, with changes free to happen.

Average users don't read the compiler specification closely enough to have any idea what is marked as unspecified or not. They see the new feature announcement which gives the basic overview (which probably won't have a giant warning about eagerness being unspecified) or they discover the feature in some codebase. Either way they play around with it a bit, and check (experimentally) to see the current behavior for things they might care about (like eager or lazy evaluation) and then assume it won't change in the future.

Reply:
> A better example is if evaluation of any element throws

As mentioned previously, evaluation order of elements is strictly defined.

Reply:
> and something which will be hard to explain to users

For the tiny set of people that run into an issue here, the situation can be explained. 

>and then assume it won't change in the future.

Then that is the consequence of behaving that way. We have made a decision balancing the needs of the ecosystem as a whole. If developers run into an issue because of that, we are ok with it on balance.

Reply:
> This could also apply to GUI scenarios like adding elements to listbox, or whatever.

It shouldn't.  You can only get the final collection if no throwing occurred during the creation of the collectino. It is all assigned to a temp.  So you either get it all, or you get nothing (since the actual code that cn reference that newly created collection doesn't run due to the exception).

Reply:
@CyrusNajmabadi  I really don't think we are talking about the same thing, and that may explain some confusion.

As best I can tell, you seem to be talking about part of the already shipped collection expression feature for building collections where iteration order of spreads when building a collection may vary relative to the evaluation of elements. If that is what you are talking about then yes, I fully agree that it is really unlikely that changes there break people in practice. 


But what I am talking about here is part of proposed specification for foreach loops over collection expressions. What I am looking at seems to say executions of the body of the foreach loop can interleave in unspecified ways with evaluating the elements of collection expression passed to the foreach loop:

> The elements of the collection are evaluated in order, and the **loop body** is executed for each element in order. The compiler *may* **evaluate subsequent elements** before **executing the loop body for preceding elements**.

Which says that  `foreach (int i in [1, myfunc()]) { foo(i); }` might do:
- evaluate 1
- run foo(1)
- evaluate myfunc() (which happens to throw)

or it might do:
- evaluate 1.
- evaluate myfunc() (which happens to throw).
- Would now try to run loop body for each, which won't happen because of the throw.

And furthermore which one it does might change in the future. That is what I have concerns about.

Reply:
I understand your concerns.  We'll def consider them during the ldm discussions on this topic.